### PR TITLE
Add structs and arrays

### PR DESCRIPTION
**Pull request by dbrtly**
Hi @wouter 

The tests fail but there are tests now.
What was your plan to enable testing with bigquery structs and arrays?

Daniel

### CODE REVIEW

**Comment by wouter** [comment]
I only made these changes to get dbt unit testing going for our specific dbt setup. I didn't plan to build it out to a full fledged feature for the dbt-unit-testing package, but if you like to do so feel free to continue this fork.

**Comment by dbrtly** [comment]
I understand. 

Does the public code you shared work for you? I added two tests but they fail for me.  

Do you have another commit that's mot publically shared?

**Comment by wouter** [comment]
Yes, these changes work for us, we use the latest version in this branch: https://github.com/wouter/dbt-unit-testing/tree/add-struct-array-udf-support-bigquery

File: macros/tests.sql [hunk 1]
```diff
@@ -78,51 +78,89 @@
...
-    select count(1) as {{ count_column }}, {{columns}} from ({{ expectations.input_values }}) as s group by {{ columns }}
+  {%- set expectations_cte -%}
+expectations as (
+    with
+    expected as (
+    {{ expectations.input_values }}
+    )
+    select
+        count(1) as {{ count_column }},
+        {{ selected_columns }}
```
Comment (NicolasGuary) [inline macros/tests.sql:113]: Hello, I tested this, but in my case, this applied `TO_JSON_STRING` twice on my column, so I replaced it by `grouped_columns` to make it work

File: macros/tests.sql [hunk 2]
```diff
@@ -150,15 +188,26 @@
   {% set actual_query = test_queries.actual_query %}
   {% set expectations_query = test_queries.expectations_query %}
   {% set test_query = test_queries.test_query %}
+  {% set udf_definitions = test_configuration.options.udf_definitions | default("udf_definitions") %}
```
Comment (NicolasGuary) [inline macros/tests.sql:191]: For some reason when testing this, the default value was a plain text string `udf_definitions` and caused the query to fail on BigQuery. Replaced by :
```
{% set udf_definitions = test_configuration.options.udf_definitions | default("") %}
```
and worked fine!

Comment (NicolasGuary) [reply macros/tests.sql:191]: Same for the other macro where you define `udf_definitions`. If that worked fine for you maybe I don't have something locally that you do, but still!

**Comment by NicolasGuary** [comment]
After a bit of changes I was able to make it work on BigQuery to test a model that performs aggregations so thank's for this!
@dbrtly is there any plans on merging this to https://github.com/EqualExperts/dbt-unit-testing anytime soon?

Also, if you plan to, I suggest we rename `convert_columns` to something maybe more explicit like `columns_to_stringify`

**Comment by dbrtly** [comment]
> After a bit of changes I was able to make it work on BigQuery to test a model that performs aggregations so thank's for this!
> 
> @dbrtly is there any plans on merging this to https://github.com/EqualExperts/dbt-unit-testing anytime soon?
> 
> 
> 
> Also, if you plan to, I suggest we rename `convert_columns` to something maybe more explicit like `columns_to_stringify`

https://github.com/EqualExperts/dbt-unit-testing/pull/193

I added 2 tests and expanded one of the macros. Stripped out the udf part to faciliate a quick code review. Let's see how that goes. Would add the udf too if it gets merged.

File: macros/tests.sql (hunk 1 above)
Comment (dbrtly) [reply macros/tests.sql:113]: Am seeing this only now. Agreed on the grouped columns edit.

**Pull request closed without merging**